File: app/src/main/java/com/mytlogos/enterprise/background/TaskManager.java
package com.mytlogos.enterprise.background;

import android.os.Looper;

import java.util.concurrent.Callable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

public class TaskManager {
    private final static TaskManager INSTANCE = new TaskManager();
    private final ExecutorService service = Executors.newCachedThreadPool();

    private TaskManager() {
        if (INSTANCE != null) {
            throw new IllegalStateException("only one instance allowed");
        }
    }

    public static TaskManager getInstance() {
        return INSTANCE;
    }

    public static <T> Future<T> runAsyncTask(Callable<T> callable) {
        return INSTANCE.service.submit(callable);
    }

    public static void runTask(Runnable task) {
        if (Looper.getMainLooper().isCurrentThread()) {
            INSTANCE.service.execute(task);
        } else {
            task.run();
        }
    }

    public static Future<?> runAsyncTask(Runnable runnable) {
        return INSTANCE.service.submit(runnable);
    }
}

File: app/src/main/java/com/mytlogos/enterprise/background/TaskManager.kt
package com.mytlogos.enterprise.background

import android.os.Looper
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.asCoroutineDispatcher
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.concurrent.Callable
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executors
import java.util.concurrent.Future
import java.util.function.Supplier
import kotlin.coroutines.coroutineContext

class TaskManager private constructor() {
    private val service = Executors.newCachedThreadPool()

    companion object {
        val instance: TaskManager = TaskManager()

        fun <T> runAsyncTask(callable: Callable<T>?): Future<T> {
            return instance.service.submit(callable)
        }

        @kotlin.jvm.JvmStatic
        fun <T> runCompletableTask(supplier: Supplier<T>?): CompletableFuture<T> {
            return CompletableFuture.supplyAsync(supplier, instance.service)
        }

        @kotlin.jvm.JvmStatic
        fun runTask(task: Runnable) {
            if (Looper.getMainLooper().isCurrentThread) {
                instance.service.execute(task)
            } else {
                task.run()
            }
        }

        @kotlin.jvm.JvmStatic
        fun runTaskSuspend(task: suspend () -> Unit) {
            CoroutineScope(instance.service.asCoroutineDispatcher()).launch {
                task()
            }
        }

        fun runAsyncTask(runnable: Runnable?): Future<*> {
            return instance.service.submit(runnable)
        }
    }
}